diff -rupN xmp-4.2.0.old/src/Makefile.am xmp-4.2.0/src/Makefile.am
--- xmp-4.2.0.old/src/Makefile.am	2023-06-15 22:11:10.000000000 +0200
+++ xmp-4.2.0/src/Makefile.am	2025-08-13 11:49:27.090746539 +0200
@@ -7,7 +7,7 @@ AM_CFLAGS   = -Wall
 bin_PROGRAMS = xmp
 
 xmp_SOURCES = \
-	commands.c common.h delay.c getopt_long.h list.h getopt_long.c info.c \
+	commands.c common.h delay.c getopt_long.h getopt_long.c info.c \
 	main.c options.c read_config.c sound.c sound.h sound_file.c sound_null.c \
 	sound_wav.c sound_aiff.c terminal.c util.c xmp_version.h
 xmp_LDADD   = ${LIBXMP_LIBS}
